src/common/mfu_flist_copy.c

```diff
@@ -2406,6 +2551,132 @@ int mfu_flist_copy(mfu_flist src_cp_list, int numpaths,
     return rc;
 }
 
+int mfu_flist_hardlink(mfu_flist src_link_list,
+        const mfu_param_path* srcpath, const mfu_param_path* destpath,
+        mfu_copy_opts_t* mfu_copy_opts)
+{
+    /* assume we'll succeed */
+    int rc = 0;
+
+    /* get our rank */
+    int rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    /* set mfu_copy options in mfu_copy_opts_t struct */
+
+    /* copy the destination path to user opts structure */
+    mfu_copy_opts->dest_path = MFU_STRDUP((*destpath).path);
+
+    /* print note about what we're doing and the amount of files/data to be moved */
+    if (rank == 0) {
+        MFU_LOG(MFU_LOG_INFO, "Linking to %s", mfu_copy_opts->dest_path);
+    }
+    mfu_flist_print_summary(src_link_list);
+
+    /* Grab a relative and actual start time for the epilogue. */
+    time(&(mfu_copy_stats.time_started));
+    mfu_copy_stats.wtime_started = MPI_Wtime();
+
+    /* Initialize statistics */
+    mfu_copy_stats.total_dirs  = 0;
+    mfu_copy_stats.total_files = 0;
+    mfu_copy_stats.total_links = 0;
+    mfu_copy_stats.total_size  = 0;
+    mfu_copy_stats.total_bytes_copied = 0;
+
+    /* Initialize file cache */
+    mfu_copy_src_cache.name = NULL;
+    mfu_copy_dst_cache.name = NULL;
+
+    /* split items in file list into sublists depending on their
+     * directory depth */
+    int levels, minlevel;
+    mfu_flist* lists;
+    mfu_flist_array_by_depth(src_link_list, &levels, &minlevel, &lists);
+
+    /* TODO: filter out files that are bigger than 0 bytes if we can't read them */
+
+    /* create directories, from top down */
+    int tmp_rc = mfu_create_directories(levels, minlevel, lists, 1,
+            srcpath, destpath, mfu_copy_opts);
+    if (tmp_rc < 0) {
+        rc = -1;
+    } else {
+        /* create hard links */
+        tmp_rc = mfu_create_hardlinks(levels, minlevel, lists,
```

Comment:
To be consistent with a normal copy, let's go ahead and try to create hardlinks even if we failed to create directories.  This is so that we may have partial success (we will still create hard links under any directories that were created).  Same for metadata.

Unless you think this is a bad idea?

Reply:
No objections, let's follow current partial success syntax.
Added a FIXME comment for this, if other ones have better idea, will improve it then.